File: src/main/java/ru/arc/stock/StockClient.java
package ru.arc.stock;

import ru.arc.util.Common;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;
import org.eclipse.jetty.util.thread.QueuedThreadPool;
import org.eclipse.jetty.websocket.api.Session;
import org.eclipse.jetty.websocket.api.WebSocketAdapter;
import org.eclipse.jetty.websocket.client.WebSocketClient;
import org.jsoup.Connection;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URL;
import java.util.*;
import java.util.concurrent.*;
import java.util.stream.Collectors;

@RequiredArgsConstructor
@Log4j2
@SuppressWarnings({"unchecked", "rawtypes"})
public class StockClient {

    private final String FINN_API_KEY;
    private final String POLY_API_KEY;

    static WebSocketClient webSocketClient;
    static volatile boolean isClosed = true;

    static Map<String, List<Double>> prices = new ConcurrentHashMap<>();

    private static final Gson gson = Common.gson;
    private static final ExecutorService service = Executors.newSingleThreadExecutor();

    public Map<String, Double> cryptoPrices() {
        String ids = StockMarket.configStocks().stream()
                .filter(stock -> stock.type == Stock.Type.CRYPTO)
                .map(ConfigStock::getSymbol)
                .collect(Collectors.joining("%2C"));
        String url = "https://api.coingecko.com/api/v3/simple/price?ids=" + ids + "&vs_currencies=usd&precision=full";
        log.debug("Fetching crypto prices from {}", url);
        try {
            HttpURLConnection connection = (HttpURLConnection) URI.create(url).toURL().openConnection();
            InputStream stream = connection.getInputStream();

            TypeToken<Map<String, Map<String, Double>>> typeToken = new TypeToken<>() {
            };
            Reader reader = new InputStreamReader(stream);
            Map<String, Map<String, Double>> map = gson.fromJson(reader, typeToken);

            Map<String, Double> prices = new HashMap<>();
            map.forEach((key, value) -> prices.put(key, value.get("usd")));
            log.debug("Fetched crypto prices: {}", prices);
            return prices;
        } catch (Exception e) {
            log.error("Could not load crypto prices", e);
            return Map.of();
        }
    }

    private static double fetchInvesting(String url) {
        try {
            Connection.Response response = Jsoup.connect(url).execute();
            if (response.statusCode() < 200 || response.statusCode() > 299) {
                // print reason for bad response
                System.out.println("Response code: " + response.statusCode());
                System.out.println(response.headers());
                System.out.println();
                System.out.println(response.body());
                return -1;
            }
            Document document = response.parse();
            Element divElement = document.select("div[data-test=instrument-price-last]").first();
            if (divElement == null) return -1;
            return Double.parseDouble(divElement.text()
                    .replace(".", "")
                    .replace(",", "."));
        } catch (Exception e) {
            log.error("Could not load price from investing.com", e);
            return -1;
        }
    }


    public void startWebSocket(Collection<String> symbols) {
        try {
            MySocket client = new MySocket();
            webSocketClient = new WebSocketClient();

            final String SERVER_URI = "wss://ws.finnhub.io?token=cn7rbt9r01qplv1e8j50cn7rbt9r01qplv1e8j5g";
            final int TIMEOUT_SECONDS = 5;

            QueuedThreadPool threadPool = new QueuedThreadPool(30, 1, 60000);
            webSocketClient.setExecutor(threadPool);
            webSocketClient.start();
            webSocketClient.connect(client, new URI(SERVER_URI));

            isClosed = false;
            service.submit(() -> {
                try {
                    if (client.latch.await(TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
                        // Send a message to the server
                        symbols.forEach(s -> {
                            try {
                                client.getSession().getRemote().sendString("{\"type\":\"subscribe\",\"symbol\":\"" + s + "\"}");
                                Thread.sleep(100);
                            } catch (Exception e) {
                                log.error("Could not send message to server", e);
                                throw new RuntimeException(e);
                            }
                        });
                        // Close the WebSocket connection
                    } else {
                        isClosed = true;
                        System.err.println("WebSocket connection could not be established within the timeout.");
                    }
                } catch (Exception e) {
                    isClosed = true;
                    //e.printStackTrace();
                }
            });
        } catch (Exception e) {
            isClosed = true;
            //e.printStackTrace();
        }
    }

    public static void stopClient() {
        if (webSocketClient != null && !isClosed) {
            try {
                webSocketClient.stop();
                isClosed = true;
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }
    }

    private Double getStockPrice(ConfigStock stock) {
        if (webSocketClient == null || !webSocketClient.isRunning() || isClosed) {
            System.out.println("Websocket is closed. Starting...");
            startWebSocket(StockMarket.stocks().stream()
                    .filter(st -> st.type == Stock.Type.STOCK)
                    .map(Stock::getSymbol)
                    .toList());
        }
        var list = prices.remove(stock.symbol);
        if (list == null) return fetchFinnhub(stock.symbol);
        OptionalDouble optional = list.stream().mapToDouble(Double::doubleValue).average();
        if (optional.isEmpty()) return fetchFinnhub(stock.symbol);
        return optional.getAsDouble();
    }

    private Double getCurrencyPrice(ConfigStock stock) {
        String url = "https://ru.investing.com/currencies/" + stock.symbol.replace("/", "-").toLowerCase();
        return fetchInvesting(url);
    }

    private Double getCommodityPrice(ConfigStock stock) {
        String url = "https://ru.investing.com/commodities/" + stock.symbol.replace("/", "-").toLowerCase();
        return fetchInvesting(url);
    }

    public Double price(ConfigStock stock) {
        return switch (stock.type) {
            case STOCK -> getStockPrice(stock);
            case CURRENCY -> getCurrencyPrice(stock);
            case COMMODITY -> getCommodityPrice(stock);
            default -> -1.0;
        };
    }

    private double fetchFinnhub(String symbol) {
        if (POLY_API_KEY == null) return -1;
        StringBuilder builder = new StringBuilder("https://finnhub.io/api/v1/quote?");
        builder.append("symbol=").append(symbol);
        builder.append("&token=").append(FINN_API_KEY);

        try {
            URL url = URI.create(builder.toString()).toURL();
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            InputStream stream = connection.getInputStream();

            Map map = gson.fromJson(new InputStreamReader(stream), Map.class);
            return (double) (map.get("c"));
        } catch (Exception e) {
            //e.printStackTrace();
            log.error("Could not load price from finnhub for {}", symbol, e);
            return -1;
        }
    }

    public double dividend(String symbol) {
        if (POLY_API_KEY == null) return 0;
        StringBuilder builder = new StringBuilder("https://api.polygon.io/v3/reference/dividends?");
        builder.append("ticker=").append(symbol);
        builder.append("&apiKey=").append(POLY_API_KEY);

        try {
            URL url = URI.create(builder.toString()).toURL();
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            InputStream stream = connection.getInputStream();

            Map map = gson.fromJson(new InputStreamReader(stream), Map.class);
            //System.out.println(map);
            return (double) ((Map<String, Object>) ((List<Object>) map.get("results")).getFirst()).get("cash_amount");
        } catch (Exception e) {
            //e.printStackTrace();
            return 0;
        }
    }

    public static class MySocket extends WebSocketAdapter {
        private final CountDownLatch latch = new CountDownLatch(1);

        @Override
        public void onWebSocketConnect(Session sess) {
            super.onWebSocketConnect(sess);
            //System.out.println("WebSocket Connected: " + sess);
            latch.countDown(); // Notify that the connection has been established
            //isClosed = false;
        }

        @Override
        public void onWebSocketText(String message) {
            super.onWebSocketText(message);
            //System.out.println("Received message from server: " + message);
            try {
                Map map = gson.fromJson(message, Map.class);
                double price = ((Number) ((Map) ((List) map.get("data")).getFirst()).get("p")).doubleValue();
                String symbol = ((String) ((Map) ((List) map.get("data")).getFirst()).get("s"));
                prices.putIfAbsent(symbol, new ArrayList<>());
                prices.get(symbol).add(price);
                //System.out.println(price+" "+symbol+" "+Thread.currentThread().getName());
            } catch (Exception e) {
                //System.out.println("Could not read data from stock socket!");
            }
        }

        @Override
        public void onWebSocketClose(int statusCode, String reason) {
            super.onWebSocketClose(statusCode, reason);
            System.out.println("WebSocket Closed: " + statusCode + " - " + reason);
            isClosed = true;
        }

        @Override
        public void onWebSocketError(Throwable cause) {
            super.onWebSocketError(cause);
            //cause.printStackTrace();
        }
    }

}

File: src/main/kotlin/ru/arc/stock/StockClient.kt
package ru.arc.stock

import com.google.gson.reflect.TypeToken
import org.jsoup.Jsoup
import ru.arc.util.Common
import ru.arc.util.Logging.debug
import ru.arc.util.Logging.error
import ru.arc.util.Logging.info
import ru.arc.util.Logging.warn
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URI
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList

@Suppress("UNCHECKED_CAST")
class StockClient internal constructor(
    finnApiKey: String?,
    polyApiKey: String?,
    private val streamFactory: StockPriceStreamFactory,
) : AutoCloseable {
    constructor(
        finnApiKey: String?,
        polyApiKey: String?,
    ) : this(
        finnApiKey = finnApiKey,
        polyApiKey = polyApiKey,
        streamFactory =
            StockPriceStreamFactory { apiKey, onPrice ->
                JettyStockPriceStream(apiKey, onPrice)
            },
    )

    private val finnApiKey = finnApiKey?.takeIf { it.isNotBlank() }
    private val polyApiKey = polyApiKey?.takeIf { it.isNotBlank() }
    private val prices = ConcurrentHashMap<String, CopyOnWriteArrayList<Double>>()
    private var stream: StockPriceStream? = null
    @Volatile
    private var lifecycleVersion = 0L

    companion object {
        private const val HTTP_TIMEOUT_MS = 10_000
        private const val HTTP_USER_AGENT = "RusCrafting-ARC/1.0"
        private val YAHOO_TICKERS = mapOf("XBR/USD" to "BZ=F")
        private val gson = Common.gson
    }

    internal val isClosed: Boolean
        @Synchronized get() = stream?.isRunning != true

    @Synchronized
    internal fun hasWebSocketClient(): Boolean = stream != null

    fun cryptoPrices(): Map<String, Double> {
        val ids =
            StockMarket.configStocks()
                .filter { it.type == Stock.Type.CRYPTO }
                .joinToString("%2C") { it.symbol }
        if (ids.isEmpty()) return emptyMap()
        val url = "https://api.coingecko.com/api/v3/simple/price?ids=$ids&vs_currencies=usd&precision=full"
        debug("Fetching crypto prices from {}", url)
        return try {
            val connection = URI.create(url).toURL().openConnection() as HttpURLConnection
            connection.connectTimeout = HTTP_TIMEOUT_MS
            connection.readTimeout = HTTP_TIMEOUT_MS
            val typeToken = object : TypeToken<Map<String, Map<String, Double>>>() {}
            val map: Map<String, Map<String, Double>> =
                InputStreamReader(connection.inputStream).use { reader ->
                    gson.fromJson(reader, typeToken)
                }
            buildMap {
                map.forEach { (key, value) -> put(key, value["usd"] ?: 0.0) }
            }.also { result ->
                debug("Fetched crypto prices: {}", result)
            }
        } catch (e: Exception) {
            error("Could not load crypto prices", e)
            emptyMap()
        }
    }

    private fun fetchInvesting(url: String): Double =
        try {
            val response = Jsoup.connect(url).timeout(HTTP_TIMEOUT_MS).execute()
            if (response.statusCode() !in 200..299) {
                warn("investing.com returned status {}", response.statusCode())
                -1.0
            } else {
                response.parse()
                    .select("div[data-test=instrument-price-last]")
                    .first()
                    ?.text()
                    ?.replace(".", "")
                    ?.replace(",", ".")
                    ?.toDoubleOrNull()
                    ?: -1.0
            }
        } catch (e: Exception) {
            error("Could not load price from investing.com", e)
            -1.0
        }

    fun startWebSocket(symbols: Collection<String>) {
        val apiKey = finnApiKey ?: return
        val normalizedSymbols =
            symbols.asSequence()
                .map(String::trim)
                .filter(String::isNotEmpty)
                .distinct()
                .toList()
        if (normalizedSymbols.isEmpty()) return
        val startVersion: Long
        val previous: StockPriceStream?
        synchronized(this) {
            if (stream?.isRunning == true) return
            previous = stream
            stream = null
            prices.clear()
            startVersion = ++lifecycleVersion
        }
        previous?.close()

        val created =
            try {
                streamFactory.create(
                    apiKey,
                    { symbol, price -> recordPrice(startVersion, symbol, price) },
                )
            } catch (e: Exception) {
                error("Could not create stock WebSocket", e)
                return
            }
        try {
            created.start(normalizedSymbols)
            val installed =
                synchronized(this) {
                    if (lifecycleVersion == startVersion && stream == null && created.isRunning) {
                        stream = created
                        true
                    } else {
                        false
                    }
                }
            if (!installed) created.close()
        } catch (e: Exception) {
            error("Could not start stock WebSocket", e)
            created.close()
        }
    }

    private fun recordPrice(version: Long, symbol: String, price: Double) {
        synchronized(this) {
            if (lifecycleVersion != version) return
            prices.computeIfAbsent(symbol) { CopyOnWriteArrayList() }.add(price)
        }
    }

    override fun close() {
        val current =
            synchronized(this) {
                lifecycleVersion++
                prices.clear()
                stream.also { stream = null }
            }
        current?.close()
    }

    private fun getStockPrice(stock: ConfigStock): Double {
        if (finnApiKey == null) return -1.0
        if (isClosed) {
            info("WebSocket is closed, reconnecting...")
            startWebSocket(
                StockMarket.stocks()
                    .filter { it.type == Stock.Type.STOCK }
                    .map { it.symbol },
            )
        }
        val buffered = prices.remove(stock.symbol) ?: return fetchFinnhub(stock.symbol)
        return buffered.stream().mapToDouble { it }.average().orElseGet { fetchFinnhub(stock.symbol) }
    }

    private fun getCurrencyPrice(stock: ConfigStock): Double {
        YAHOO_TICKERS[stock.symbol]?.let { ticker ->
            return fetchYahooFinance(ticker)
        }
        val url = "https://ru.investing.com/currencies/${stock.symbol.replace("/", "-").lowercase()}"
        return fetchInvesting(url)
    }

    private fun getCommodityPrice(stock: ConfigStock): Double {
        val url = "https://ru.investing.com/commodities/${stock.symbol.replace("/", "-").lowercase()}"
        return fetchInvesting(url)
    }

    fun price(stock: ConfigStock): Double =
        when (stock.type) {
            Stock.Type.STOCK -> getStockPrice(stock)
            Stock.Type.CURRENCY -> getCurrencyPrice(stock)
            Stock.Type.COMMODITY -> getCommodityPrice(stock)
            else -> -1.0
        }

    private fun fetchYahooFinance(ticker: String): Double {
        val encodedTicker = URLEncoder.encode(ticker, StandardCharsets.UTF_8)
        val url = "https://query1.finance.yahoo.com/v8/finance/chart/$encodedTicker?interval=1d&range=1d"
        return try {
            val connection = URI.create(url).toURL().openConnection() as HttpURLConnection
            connection.connectTimeout = HTTP_TIMEOUT_MS
            connection.readTimeout = HTTP_TIMEOUT_MS
            connection.setRequestProperty("User-Agent", HTTP_USER_AGENT)
            val body =
                InputStreamReader(connection.inputStream).use { reader ->
                    reader.readText()
                }
            parseYahooRegularMarketPrice(body) ?: run {
                warn("Yahoo Finance returned no market price for {}", ticker)
                -1.0
            }
        } catch (e: Exception) {
            error("Could not load price from Yahoo Finance for {}", ticker, e)
            -1.0
        }
    }

    private fun fetchFinnhub(symbol: String): Double {
        if (finnApiKey == null) return -1.0
        val request = finnhubQuoteRequest(symbol, finnApiKey)
        return try {
            val connection = request.uri.toURL().openConnection() as HttpURLConnection
            connection.connectTimeout = HTTP_TIMEOUT_MS
            connection.readTimeout = HTTP_TIMEOUT_MS
            request.headers.forEach(connection::setRequestProperty)
            val map =
                InputStreamReader(connection.inputStream).use { reader ->
                    gson.fromJson(reader, Map::class.java) as Map<String, Any>
                }
            (map["c"] as Number).toDouble()
        } catch (e: Exception) {
            error("Could not load price from finnhub for {}", symbol, e)
            -1.0
        }
    }

    fun dividend(symbol: String): Double {
        if (polyApiKey == null) return 0.0
        val url = "https://api.polygon.io/v3/reference/dividends?ticker=$symbol&apiKey=$polyApiKey"
        return try {
            val connection = URI.create(url).toURL().openConnection() as HttpURLConnection
            connection.connectTimeout = HTTP_TIMEOUT_MS
            connection.readTimeout = HTTP_TIMEOUT_MS
            val map =
                InputStreamReader(connection.inputStream).use { reader ->
                    gson.fromJson(reader, Map::class.java) as Map<String, Any>
                }
            val results = map["results"] as? List<*> ?: return 0.0
            val first = results.firstOrNull() as? Map<*, *> ?: return 0.0
            (first["cash_amount"] as? Number)?.toDouble() ?: 0.0
        } catch (_: Exception) {
            0.0
        }
    }
}

internal data class FinnhubQuoteRequest(
    val uri: URI,
    val headers: Map<String, String>,
)

internal fun finnhubQuoteRequest(
    symbol: String,
    apiKey: String,
): FinnhubQuoteRequest =
    FinnhubQuoteRequest(
        uri =
            URI.create(
                "https://finnhub.io/api/v1/quote?symbol=" +
                    URLEncoder.encode(symbol, StandardCharsets.UTF_8),
            ),
        headers = mapOf("X-Finnhub-Token" to apiKey),
    )

internal fun parseYahooRegularMarketPrice(payload: String): Double? =
    runCatching {
        val root = Common.gson.fromJson(payload, Map::class.java) ?: return@runCatching null
        val chart = root["chart"] as? Map<*, *> ?: return@runCatching null
        val result = (chart["result"] as? List<*>)?.firstOrNull() as? Map<*, *> ?: return@runCatching null
        val meta = result["meta"] as? Map<*, *> ?: return@runCatching null
        (meta["regularMarketPrice"] as? Number)?.toDouble()
    }.getOrNull()
